x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/xpack/searchablesnapshots/SearchableSnapshotEnableAllocationDecider.java

```diff
@@ -0,0 +1,100 @@
+                 + "] is not valid");
+        }
+        return allocation;
+    }
+
+    private volatile EnableAllocationDecider.Allocation enableAllocation;
+    private volatile EnableAllocationDecider.Allocation primariesAllocation;
+
+    public SearchableSnapshotEnableAllocationDecider(Settings settings, ClusterSettings clusterSettings) {
+        this.enableAllocation = EnableAllocationDecider.CLUSTER_ROUTING_ALLOCATION_ENABLE_SETTING.get(settings);
+        this.primariesAllocation = SEARCHABLE_SNAPSHOTS_ALLOCATION_ENABLE_PRIMARIES_SETTING.get(settings);
+        assertSettingValue();
+        clusterSettings.addSettingsUpdateConsumer(EnableAllocationDecider.CLUSTER_ROUTING_ALLOCATION_ENABLE_SETTING,
+            this::setEnableAllocation);
+        clusterSettings.addSettingsUpdateConsumer(SEARCHABLE_SNAPSHOTS_ALLOCATION_ENABLE_PRIMARIES_SETTING,
+            this::setPrimariesAllocation);
+    }
+
+    public void assertSettingValue() {
+        assert this.primariesAllocation != EnableAllocationDecider.Allocation.ALL && this.primariesAllocation != EnableAllocationDecider.Allocation.NEW_PRIMARIES;
+    }
+
+    private void setPrimariesAllocation(EnableAllocationDecider.Allocation allocation) {
+        this.primariesAllocation = allocation;
+        assertSettingValue();
+    }
+
+    private void setEnableAllocation(EnableAllocationDecider.Allocation allocation) {
+        this.enableAllocation = allocation;
+    }
+
+    @Override
+    public Decision canAllocate(ShardRouting shardRouting, RoutingNode node, RoutingAllocation allocation) {
+        return canAllocate(shardRouting, allocation);
+    }
+
+    @Override
+    public Decision canAllocate(ShardRouting shardRouting, RoutingAllocation allocation) {
+        final IndexMetadata indexMetadata = allocation.metadata().getIndexSafe(shardRouting.index());
+        if (SearchableSnapshotsConstants.isSearchableSnapshotStore(indexMetadata.getSettings())) {
+            EnableAllocationDecider.Allocation enableAllocation = this.enableAllocation;
+            EnableAllocationDecider.Allocation primariesAllocation = this.primariesAllocation;
+            if (enableAllocation == EnableAllocationDecider.Allocation.PRIMARIES) {
+                if (primariesAllocation == EnableAllocationDecider.Allocation.NONE) {
+                    return allocation.decision(Decision.NO, NAME,
+                        "no allocations of searchable snapshots allowed due to [%s=%s] and [%s=%s]",
+                        EnableAllocationDecider.CLUSTER_ROUTING_ALLOCATION_ENABLE_SETTING.getKey(),
+                        enableAllocation,
+                        SEARCHABLE_SNAPSHOTS_ALLOCATION_ENABLE_PRIMARIES_SETTING.getKey(),
+                        primariesAllocation
+                    );
+                } else {
+                    return allocation.decision(Decision.YES, NAME, "decider relies on generic enable decider");
```

Comment:
I think this message is wrong, the decider return `YES` here because `xpack.searchable.snapshot.allocation.enable.primaries: primaries` permits it.